internal/events/message.go

```diff
@@ -0,0 +1,85 @@
+package events
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/stellar/stellar-disbursement-platform-backend/stellar-multitenant/pkg/tenant"
+)
+
+var (
+	ErrTopicRequired    = errors.New("message topic is required")
+	ErrKeyRequired      = errors.New("message key is required")
+	ErrTenantIDRequired = errors.New("message tenant ID is required")
+	ErrTypeRequired     = errors.New("message type is required")
+	ErrDataRequired     = errors.New("message data is required")
+)
+
+type Message struct {
+	Topic    string  `json:"topic"`
+	Key      string  `json:"key"`
+	TenantID string  `json:"tenant_id"`
+	Type     string  `json:"type"`
+	Data     any     `json:"data"`
+	Errors   []Error `json:"errors,omitempty"`
+}
+
+type Error struct {
+	FailedAt     time.Time `json:"failed_at"`
+	ErrorMessage string    `json:"error_message"`
+}
+
+// NewMessage returns a new message with values passed by parameters. It also parses the `TenantID` from the context and inject it into the message.
+// Returns error if the tenant is not found in the context.
+func NewMessage(ctx context.Context, topic, key, messageType string, data any) (*Message, error) {
+	tnt, err := tenant.GetTenantFromContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting tenant from context: %w", err)
+	}
+
+	return &Message{
+		Topic:    topic,
+		Key:      key,
+		TenantID: tnt.ID,
+		Type:     messageType,
+		Data:     data,
+	}, nil
+}
+
+func (m *Message) String() string {
+	return fmt.Sprintf("Message{Topic: %s, Key: %s, Type: %s, TenantID: %s, Data: %v}", m.Topic, m.Key, m.Type, m.TenantID, m.Data)
+}
```

Comment:
Yeah so the issue is that `RecordError` needs to be done on a pointer receiver, but this one should ideally be done on the value receiver and in golang it's not recommended to mix both. I'll try to figure something out.

Reply:
I made the change in 0c3d27a98b51c393120e37681afdc0dda3c8a54c

Reply:
TY!

Here's a playground showing the effect of adding `.String()` to the value or the pointer: https://go.dev/play/p/TEyoAYP-Frz

Reply:
I believe the biggest issue with adding some methods to the pointer and others to the value is when you're talking about interfaces and interface implementations. For instance, if you had an interface:
```go
type FooBarInterface interface {
    func Foo()
    func Bar()
}
```

The following struct would not be implementing the interface correctly, because neither `FooBar{}` nor `&FooBar{}` match the desired interface:
```go
type FooBar struct {}

func (fb *FooBar) Foo() {
    fmt.Println("FOO")
}

func (fb FooBar) Bar() { // <----- this is not recommended, because the pointer implements one of the interface methods while the value implements the other.
    fmt.Println("BAR")
}

````